### ISSUE TITLE
Mixed IT-Electronics deployments and state names

### ISSUE THREAD

**Issue by termcap** [NagiosEnterprises/nagioscore#359]
Hi,

As more and more electronics devices are getting IP enabled, it is becoming possible to deploy Nagios(via SNMP monitoring) in mixed IT-Electronics environments as opposed to the old purely IT setups.

But this comes with a set of challenges. The WARNING-CRITICAL-UNKNOWN-OK states work well with purely IT setups and devices, but implementing Nagios on Electronics environments require more states, for instance MINOR-MAJOR-CRITICAL-UNKNOWN-OK would be the states that people who work with electronic devices are more familiar and comfortable with.  Maybe this could be done by allowing users to set the state names to whatever they prefer.

I understand that this is a big change as state names are part of the Nagios code logic, but I would urge that further releases of Nagios start making space for such a change or else Nagios and by extension open source will lose out on a large segment of the monitoring market.

This issue was first raised at -> https://support.nagios.com/forum/viewtopic.php?f=7&t=43301

Thanks,
termcap.

**Comment by ericloyd** [comment]
Agreed.  In fact, to take it a step further, I'd suggest that return codes themselves be parameterized such that zero is always "OK" but one can specify any range of return codes for severity levels ranging from "just below OK" all the way to "OMG it's on FIRE!"

Example:

0 = OK
1 = Working, but still want to know about it
2 = Working, but will fail soon
3 = Not working as intended, but still working
4 = Not working as intended and someone should know
5 = Broken; please inform NOC personnel immediately
6 = On fire; call fire department

This combines the best of all worlds of full customization.

**Comment by rcknag** [comment]
+1

I think the most control from a managing stand point would come from custom rules. _every_ org. will want to do this a different way, which is completely okay, but why not make the thresholds dynamic?

I agree 0 should remain OK, but 1-9 could be customized based on organization. This would take a complex setup, but allow the customization of 1-9 with actions assigned to each threshold. The other case to point is there might need to be a 'generic' wrapper script used to re-write current exit codes that exist, to bring forth growth for what's to come.

This would really help with assigning severity across the board too.

**Comment by jfrickson** [comment]
We might be able to do something, but it will be a pretty big job. I'm tentatively scheduling this for Core 5.

**Comment by termcap** [comment]
Thanks for considering the suggestion.

**Comment by hedenface** [comment]
I'll +1 the whole thing - this is the kind of contribution that really makes a product take a step past great into awesome. Having the control to define your current states would take the flexibility of Nagios to another level.

Thinking out loud (typing?) here, but when John said that it "will be a pretty big job" he wasn't kidding. This goes beyond Core into the Plugins (not just ours) and the Plugin development guidelines and even beyond that.

We'd have to come up with some kind of way for the Core to process the output from the plugins (or change all of them, developed by all those plugin developers worldwide, simultaneously, while still maintaining backwards compatibility - likely not going to happen) along with the return codes from those plugins.

Again, **this is an amazing idea**, but not only will we have to review it - but we'll need a serious plan of action. So, having said that, if anyone has any specific idea of how to develop this (better than just "we should do X thing"), then I am definitely open to hearing (seeing?) it. Otherwise, you'll have to let me put a plan together to review at the next (or more likely the one after or even after that) internal meeting we hold to go over stuff like this.

I **personally** am OK (`./check_prefs -A "major breaking changes" --who=hedenface`) with major breaking changes if they're for the sake of a Good Thing. But we'd really need to hear community feedback on the subject (in regards to the breaking changes likely required, i.e.: maybe your custom plugins, or the plugins you downloaded from whatever source, etc. stop working entirely - or they didn't adhere to the guidelines [oh geez does this even happen?] and the conversion we have in place doesn't work properly).

I'll keep this updated with any news I have regarding this particular topic. Thanks guys!

**Comment by ericloyd** [comment]
Three words:  Negative Exit Statuses

Everything else can stay the same, but plugins that exit with status less than zero can take advantage of the new parameterization (which obviously, hasn't been written yet).

**Comment by hedenface** [comment]
A haiku:

Wish that was my thought
A great idea for certain
Lets hope it's a go

Seriously, I hadn't even considered that. Any other additions are welcome
as well :)

**Comment by ericloyd** [comment]
That's why I drive a 2013 Honda!
